### PR TITLE
[Core] Sglang integration v0

### PR DESCRIPTION
**Pull request by Oasis-Git**
Core implementation for Sglang Integration

Note: The branch is not ready to merge.

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to LMCache! Before submitting the pull request, please ensure the PR meets the following criteria. This helps us maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Please try to classify PRs for easy understanding of the type of changes. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Core]</code> for changes in the core LMCache logic (e.g., <code>LMCacheEngine</code>, <code>Backend</code> etc.)</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>Pass all linter checks. Please use <a href="https://github.com/LMCache/LMCache/blob/dev/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li> Please include sufficient tests to ensure the change is stay correct and robust. This includes both unit tests and integration tests.</li>
</ul>

<h3>What to Expect for the Reviews</h3>

We aim to address all PRs in a timely manner. If no one reviews your PR within 5 days, please @-mention one of KuntaiDu, ApostaC or YaoJiayi.

### CODE REVIEW

File: examples/sglang_interface/config.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
+# The size of the chunk as an integer
+# (set to 256 by default)
+chunk_size: 256
+
+# The local KV cache device to use (set to "cuda" by default)
+# Possible values: "cpu", "cuda", "file://local_disk/"
+local_disk: file:///data/user_data/yuweia/integration/
```
Comment (YaoJiayi) [inline examples/sglang_interface/config.yaml]: Can we change this to a relative path?

File: benchmarks/sglang-integration/config.yaml [hunk 2]
```diff
@@ -0,0 +1,16 @@
+# The size of the chunk as an integer
+# (set to 256 by default)
+chunk_size: 256
+
+# The local KV cache device to use (set to "cuda" by default)
+# Possible values: "cpu", "cuda", "file://local_disk/"
+local_disk: file:///data/user_data/yuweia/integration/
```
Comment (YaoJiayi) [inline benchmarks/sglang-integration/config.yaml]: Can we change this to a relative path?

File: benchmarks/sglang-integration/config.yaml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+# The maximum size of the local KV cache as an integer (GB)
+# Set to 5 by default
+max_local_disk_size: 100
+
+max_local_cpu_size: 25
```
Comment (YaoJiayi) [inline benchmarks/sglang-integration/config.yaml:13]: IIUC, is this 25GB cpu gonna be conflicted with the CPU memory allocated by SGLang itself?

Comment (YaoJiayi) [inline benchmarks/sglang-integration/config.yaml:13]: IIUC, is this 25GB cpu gonna be conflicted with the CPU memory allocated by SGLang itself?

File: format.sh [hunk 4]
```diff
@@ -109,10 +109,10 @@ fi
 echo 'lmcache yapf: Done'
 
 # Run mypy
-echo 'lmcache mypy:'
-mypy
-mypy lmcache --follow-imports skip
-echo 'lmcache mypy: Done'
+# echo 'lmcache mypy:'
```
Comment (YaoJiayi) [inline format.sh:112]: I don't think we should remove this?

File: examples/sglang_interface/README.md [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+lmcache_retrieve_kv_hash(engine, hash_, kv_caches, retrieve_status)
+```
+**Parameters:**
+
+- `engine` (`LMCacheEngine`): Cache engine.
+- `hash_` (`List[CacheEngineKey]`): Hash keys for retrieval.
+- `kv_caches` (`torch.Tensor` or `List[torch.Tensor]`): Buffer to fill.
+- `retrieve_status` (`List[RetrieveStatus]`): Retrieval status per chunk.
+
+**Returns:** None
```
Comment (YaoJiayi) [inline examples/sglang_interface/README.md]: Why do we need two sets of interfaces? One is based on `token_ids` while the other one is based on `hash`

File: examples/sglang_interface/sglang_adapter.py [hunk 6]
```diff
@@ -0,0 +1,166 @@
...
+    kv_pool_size = 10000
+    seq_len = 512
+    engine = init_lmcache_engine(model_config, rank, world_size,
+                                 tensor_parallel_size)
+
+    # Generate a random tensor to simulate tokens, 512 tokens
+    tokens = torch.randint(0, 10000, (seq_len, )).cuda()
+
+    # Generate a kv cache pool same as hierarchical kv cache
+    kv_cache = torch.randn(2, model_config.num_hidden_layers, kv_pool_size,
```
Comment (YaoJiayi) [inline examples/sglang_interface/sglang_adapter.py:32]: Is this kv cache pool allocated by SGLang?

File: examples/sglang_interface/sglang_adapter.py [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+                           model_config.get_num_kv_heads(tensor_parallel_size),
+                           model_config.head_dim)
+
+    indices = torch.tensor([i for i in range(seq_len)])
+
+    store_status = torch.full((seq_len // engine.config.chunk_size, ),
+                              StoreStatus.FAIL,
+                              dtype=torch.int32)
+    hash_keys = lmcache_store_kv(engine, tokens, kv_cache[:, :, indices],
```
Comment (YaoJiayi) [inline examples/sglang_interface/sglang_adapter.py:41]: Can we reuse the same interface design as vllm? For example, instead of updating the `store_status` internally, can we return a `mask` that indicates which token should be stored?

Comment (YaoJiayi) [reply examples/sglang_interface/sglang_adapter.py:41]: The original semantics of `store_status` is just a indication of whether an entire request should be stored in the current iteration. For example, we do not want to store the KV cache generated during the decode iterations.

File: lmcache/experimental/cache_engine.py [hunk 8]
```diff
@@ -89,7 +96,7 @@ def __init__(
     def store(self,
               tokens: torch.Tensor,
               mask: Optional[torch.Tensor] = None,
-              **kwargs) -> None:
+              **kwargs) -> List[CacheEngineKey]:
```
Comment (YaoJiayi) [inline lmcache/experimental/cache_engine.py:99]: A general comment: I think the current interface is already very flexible due to `the existence of `**kwargs`. Maybe we don't have change the return type here? Otherwise, it might cause compatibility issues with vllm?

File: lmcache/experimental/cache_engine.py [hunk 9]
```diff
@@ -197,14 +216,81 @@ def retrieve(self,
             # cpu tensor for the sake of performance.
             # For example, disk->gpu is faster than disk->cpu->gpu.
             # RDMA is another example.
-
-            self.gpu_connector.to_gpu(memory_obj, start, end, **kwargs)
+            if self.gpu_connector is not None:
+                self.gpu_connector.to_gpu(memory_obj, start, end, **kwargs)
+            elif self.dram_connector is not None:
+                self.dram_connector.to_dram(memory_obj, start, end, **kwargs)
```
Comment (YaoJiayi) [inline lmcache/experimental/cache_engine.py:218]: We can use a single variable name here. No need to check `gpu_connector` or `dram_connector` everytime.

File: lmcache/experimental/cache_engine.py [hunk 10]
```diff
@@ -240,6 +326,18 @@ def lookup(
                 return start
         return end
 
+    def get_hash(
```
Comment (YaoJiayi) [inline lmcache/experimental/cache_engine.py:325]: It's a bit wiered to have a function `get_hash()` here?

File: lmcache/experimental/memory_management.py [hunk 11]
```diff
@@ -406,9 +406,9 @@ def allocate(
             if block.size >= aligned_size:
                 break
         else:
-            logger.warning(f"Failed to allocate memory for "
-                           f"tensor({shape}, {dtype}) because "
-                           "no memory is available")
+            # logger.warning(f"Failed to allocate memory for "
+            #                f"tensor({shape}, {dtype}) because "
+            #                "no memory is available")
```
Comment (YaoJiayi) [inline lmcache/experimental/memory_management.py:412]: Why removing?

File: lmcache/experimental/storage_backend/storage_manager.py [hunk 12]
```diff
@@ -88,13 +88,15 @@ def allocate(
...
             memory_obj = self.memory_allocator.allocate(shape, dtype)
-            logger.debug("Evicting 1 chunk from hot cache")
+            # logger.debug("Evicting 1 chunk from hot cache")
             if memory_obj is not None:
                 break
             # TODO(Jiayi): move this before the loop
             # In this way, we don't need to do eviction for big objects
             # TODO(Jiayi): the following code is hacky, please refactor
             if self.memory_allocator.pin_allocator.num_active_allocations == 0:
+                logger.debug("No active allocations in pin allocator, "
```
Comment (YaoJiayi) [inline lmcache/experimental/storage_backend/storage_manager.py:98]: Why are we skippng eviction here?

File: lmcache/integration/sglang/sglang_adapter.py [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+    :param hash_: The CacheEngineKey list for the retrieve block.
+    :param kv_caches: The KV caches to be retrieved in.
+    :param retrieve_status: The status of the retrieve operation.
+    :return: None
+    """
+    if len(retrieve_status) == 0:
+        retrieve_status = [RetrieveStatus.FAIL] * len(hash_)
+    engine.hash_retrieve(hash_,
+                         kvcaches=kv_caches,
+                         retrieve_status=retrieve_status)
```
Comment (YaoJiayi) [inline lmcache/integration/sglang/sglang_adapter.py:201]: I feel there should be should_store and should_retrieve interfaces as well?

File: lmcache/experimental/storage_backend/local_disk_backend.py [hunk 14]
```diff
@@ -260,3 +266,14 @@ def close(self) -> None:
...
+        # store the self.dict to disk
+        logger.debug(
+            "Closing LocalDiskBackend and storing the self.dict to disk.")
+        metadata_file_path = os.path.join(self.path, 'metadata_dict.pkl')
+        try:
+            with open(metadata_file_path, 'wb') as f:
+                pickle.dump(self.dict, f)
+        except Exception as e:
+            logger.debug("Failed to store metadata dictionary to disk.")
+            logger.debug(e)
```
Comment (YaoJiayi) [inline lmcache/experimental/storage_backend/local_disk_backend.py:273]: I feel we do not have to add persistent disk caching for now? I think it should be in another PR because both vllm and sgl should have this. In addition, this should be turned on/off with user config.

**Review by YaoJiayi — COMMENTED** [review]
Hey @Oasis-Git, thanks for the huge effort here! I left some comments:)

File: lmcache/integration/sglang/sglang_adapter.py [hunk 15]
```diff
@@ -0,0 +1,196 @@
+from enum import IntEnum
+from typing import List, Optional, Union
+
+import torch
+from sglang.srt.configs.model_config import ModelConfig
```
Comment (YaoJiayi) [inline lmcache/integration/sglang/sglang_adapter.py:5]: Will this cause failure if user only installed vllm?

File: benchmarks/sglang-integration/config.yaml (hunk 3 above)
Comment (Oasis-Git) [reply benchmarks/sglang-integration/config.yaml:13]: Yes, in our experiment 15GB is a good settings

File: format.sh (hunk 4 above)
Comment (Oasis-Git) [reply format.sh:112]: The mypy test may fail locally

File: examples/sglang_interface/README.md (hunk 5 above)
Comment (Oasis-Git) [reply examples/sglang_interface/README.md]: This is designed by the requirement of sglang design. For each block affiliated to one tree node in radix tree, it will store the hash for this block and retrieve

File: examples/sglang_interface/sglang_adapter.py (hunk 6 above)
Comment (Oasis-Git) [reply examples/sglang_interface/sglang_adapter.py:32]: Yes, it is similar to the one sglang sets their kvpool

File: lmcache/integration/sglang/sglang_adapter.py (hunk 15 above)
Comment (Oasis-Git) [reply lmcache/integration/sglang/sglang_adapter.py:5]: I do not think so since they do not need to include this file

File: examples/sglang_interface/sglang_adapter.py (hunk 7 above)
Comment (Oasis-Git) [reply examples/sglang_interface/sglang_adapter.py:41]: The requirement from sglang side is that they hope the whole operation to be atomic

File: lmcache/experimental/cache_engine.py (hunk 10 above)
Comment (Oasis-Git) [reply lmcache/experimental/cache_engine.py:325]: For the use of sglang

File: lmcache/experimental/memory_management.py (hunk 11 above)
Comment (Oasis-Git) [reply lmcache/experimental/memory_management.py:412]: I will recover it

File: lmcache/integration/sglang/sglang_adapter.py (hunk 15 above)
Comment (KuntaiDu) [reply lmcache/integration/sglang/sglang_adapter.py:5]: vLLM integration does not include this file so it's fine.

**Comment by wqlxx** [comment]
@Oasis-Git Do you have implemented the code in sglang to test this pr for lmcache？

**Comment by Oasis-Git** [comment]
If you are interested in the sgl integration, we move the PR to #869 . This PR will close.
